chromium/rules.js

```diff
@@ -340,6 +350,70 @@ RuleSets.prototype = {
     util.log(util.INFO, 'done removing rule');
     return true;
   },
+  
+  /**
+  * Load stored user rules
+  **/
+  getStoredUserRules: function() {
+    const oldUserRuleString = store.localStorage.getItem(this.USER_RULE_KEY);
+    let oldUserRules = [];
+    if (oldUserRuleString) {
+      oldUserRules = JSON.parse(oldUserRuleString);
+    }
+    return oldUserRules;
+  },
+
+  /**
+  * Load all stored user rules
+  */
+  loadStoredUserRules: function() {
+    const user_rules = this.getStoredUserRules();
+    for (let user_rule of user_rules) {
+      this.addUserRule(user_rule);
+    }
+    util.log(util.INFO, 'loaded ' + user_rules.length + ' stored user rules');
+  },
+
+  /**
+  * Adds a new user rule
+  * @param params: params defining the rule
+  * @param cb: Callback to call after success/fail
+  * */
+  addNewRuleAndStore: function(params) {
+    return new Promise(resolve => {
+      if (this.addUserRule(params)) {
+        // If we successfully added the user rule, save it in local
+        // storage so it's automatically applied when the extension is
+        // reloaded.
+        var oldUserRules = this.getStoredUserRules();
+        // TODO: there's a race condition here, if this code is ever executed from multiple
+        // client windows in different event loops.
+        oldUserRules.push(params);
+        // TODO: can we exceed the max size for storage?
+        store.localStorage.setItem(this.USER_RULE_KEY, JSON.stringify(oldUserRules));
+        resolve(true);
+      } else {
+        resolve(false);
```

Comment:
It looks like `addNewRule` (the function that this replaced), isn't even asynchronous, so it should have probably just returned a value, instead of using a promise. That's a little out of scope of this PR I think though.

Maybe the function used to be asynchronous because it used `chrome.storage.*`? Any idea?

Continuing this tangent, why do we use both `localStorage` AND `chrome.storage`? Is which data we store where documented somewhere?

Reply:
I think we should change to `chrome.storage` for everything, but I wasn't involved in the initial writing of the WebExtension so there might be some historical reason for this.  @semenko may know better.

In the meantime I've changed this to a synchronous method.